### ISSUE TITLE
Error appears in CI only for importing SVG in tsx file

### ISSUE THREAD

**Issue by Andrew-Chen-Wang** [oxc-project/oxc#20764]
Error: Cannot find module '@website/app/moon.svg' or its corresponding type declarations.

I'm in a NextJS monorepo. The import looks like:

```
import Moon from "@website/app/moon.svg"
```

in a .tsx file

This error isn't reported locally, but it appears in CI.

Ok, so according to Claude:

> The issue is clear. Locally you have .next/dev/types/routes.d.ts (generated by running the dev server), but on CI that       
  directory doesn't exist. The next-env.d.ts imports from it, and when that fails, the SVG type declarations from              
  next/image-types/global don't get loaded properly by oxlint's type checker.

then it did:

```
⏺ Write(apps/website/src/assets.d.ts)                                                                                          
  ⎿  Wrote 8 lines to apps/website/src/assets.d.ts                                                                             
      1 // Declare static asset modules for oxlint type checking in CI                                                         
      2 // where .next/dev/types don't exist and next-env.d.ts can't                                                           
      3 // load next/image-types/global declarations.                                                                          
      4                                                           
      5 declare module "*.svg" {                                                                                               
      6   const content: any                                      
      7   export default content                        
      8 }
```

I simply combined my oxlint with my regular build + test pipeline.

**Comment by camc314** [comment]
Hi,

can you share:
 - the exact error you're seeing and where it's coming from - e.g. in your editor, CLI ect
 - your oxlint config
 - tsconfig
 - if you're in a monorepo ect

This will help us to be able to track where the problem is coming from.

**Comment by Andrew-Chen-Wang** [comment]
- It's a pnpm monorepo. `apps/website` contains `@website` which is my nextjs project

You can try this using my template https://github.com/Andrew-Chen-Wang/template-nextjs You'll have to add oxlint though. You'll also need to generate a svg file, then import it into a tsx file to reproduce.

Oxlint:
```ts
import { defineConfig } from "oxlint"

export default defineConfig({
  // https://oxc.rs/docs/guide/usage/linter/plugins.html
  plugins: [
    "eslint",
    // "react",
    "typescript",
    "nextjs",
    "node",
    "vitest",
    "jsdoc",
    "import",
    // "unicorn"
  ],
  categories: {
    correctness: "error",
    suspicious: "warn",
    perf: "warn",
  },
  options: {
    typeAware: true,
    typeCheck: true,
  },
  env: {
    node: true,
    builtin: true,
  },
  ignorePatterns: [
    "**/.next/**",
    "private_notes/**",
    "**/.config/**",
    "apps/website/postcss.config.js",
    "apps/website/next.config.ts",
    "packages/search/training/**",
    "apps/website/src/services/client/**",
  ],
  rules: {
    "@typescript-eslint/restrict-plus-operands": [
      "error",
      {
        allowAny: false,
        allowBoolean: false,
        allowNullish: false,
        allowNumberAndString: false,
        allowRegExp: false,
      },
    ],
    "@typescript-eslint/restrict-template-expressions": [
      "error",
      {
        allowNumber: true,
        allowBoolean: true,
      },
    ],
    "@typescript-eslint/return-await": ["error", "error-handling-correctness-only"],
    "no-unsafe-type-assertion": "off",
    "no-shadow": "off",
    "no-unused-vars": "off",
    "no-unused-expressions": "off",
    "no-await-in-loop": "off",
    "no-named-as-default-member": "off",
    "no-img-element": "allow",
    "no-named-as-default": "off",
    "no-unassigned-import": ["error", { allow: ["**/*.css"] }],
    "react/react-in-jsx-scope": "off",
  },
  overrides: [
    {
      files: ["**/*.ts", "**/*.tsx", "**/*.mts", "**/*.cts"],
      rules: {
        "constructor-super": "off",
        "no-class-assign": "off",
        "no-const-assign": "off",
        "no-dupe-class-members": "off",
        "no-dupe-keys": "off",
        "no-func-assign": "off",
        "no-import-assign": "off",
        "no-new-native-nonconstructor": "off",
        "no-obj-calls": "off",
        "no-redeclare": "off",
        "no-setter-return": "off",
        "no-this-before-super": "off",
        "no-unsafe-negation": "off",
        "no-var": "error",
        "no-with": "off",
        "prefer-const": "error",
        "prefer-rest-params": "error",
        "prefer-spread": "error",
      },
    },
  ],
})

```

You can disable all the plugins; initially when I ran this in CI, there was only the typescript plugin.

tsconfig:

```json
{
  "compilerOptions": {
    "lib": ["dom", "dom.iterable", "esnext"],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "noEmit": true,
    "target": "ESNext",
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "bundler",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "allowImportingTsExtensions": false,
    "incremental": true,
    "composite": true,
    "plugins": [
      {
        "name": "next"
      }
    ],
    "paths": {
      "@website/*": ["./src/*"],
      "@lib/dao/*": ["../../lib/typescript/dao/src/*"]
    }
  },
  "include": ["next-env.d.ts", "**/*.ts", "**/*.tsx", ".next/types/**/*.ts"],
  "exclude": ["node_modules"],
  "references": [{ "path": "../../lib/typescript/dao" }]
}

```

The original file(s) that this was reported from was a page.tsx file, server component in NextJS. I have an SVG file that is imported into the page.tsx file

```tsx
import Moon from "@website/app/moon.svg"
```
This was the only error I saw in CI: `Error: Cannot find module '@website/app/moon.svg' or its corresponding type declarations.` Again, I didn't see this locally because `.next/dev` was already there. The fix I added for myself was to run `npm run build`/`next build` first and then run oxlint, and that fixed it.

<img width="634" height="481" alt="Image" src="https://github.com/user-attachments/assets/8370887a-f89f-476b-9a2f-6659f62b93d1" />

Claude Code says the debug logs (when adding the OXC_DEBUG var) didn't show anything; it was just a guess

> You're right to call that out. The debug logs don't actually mention next-env.d.ts or .next/dev/types/routes.d.ts as the
  cause. That was my hypothesis for why it works locally but not on CI — it wasn't stated in the logs.
                                                                                                                               
>  What the logs actually show is simpler: oxlint resolves @website/app/moon.svg to the correct file on disk, but then fails    
  because there's no declare module "*.svg" type declaration available. The fix is the same either way — the standalone        
  assets.d.ts provides that declaration — but I should have been upfront that the next-env.d.ts cascade theory was speculation,
   not something from the logs

but i mean running the build first did work so...

**Comment by camc314** [comment]
Hi, I tried to reproduce here, but I was unable to.

https://github.com/camc314/template-nextjs/commit/357fc9e9d82afebb919998303c1743be8ab68581

I think claude code is correct here, typescript is relying on some files that are generated by Next to perform typechecking, if those files are not there, then typechecking fails.

The solution is to:
1. build before typechecking
2. manually declare your svg files in a declaration file so typescript knows how to interpret them

Thanks

**Comment by Andrew-Chen-Wang** [comment]
I suggest for reproducing, you should try deleting the `.next` folder and running oxlint again.

I suppose this isn't really a bug, but perhaps there could be better hints? I'm not sure if that's possible

**Comment by camc314** [comment]
Not sure how we can really give better hints here - we'd have exactly the same errors as typescript if you ran typescript's type check prior to building.

> I suggest for reproducing, you should try deleting the .next folder and running oxlint again.

I still couldn't reproduce - I see this: in /template-nextjs/node_modules/next/image-types/global.d.ts which i think it providing the types - not sure why you don't have it.


```
declare module '*.svg' {
  /**
   * Use `any` to avoid conflicts with
   * `@svgr/webpack` plugin or
   * `babel-plugin-inline-react-svg` plugin.
   */
  const content: any

  export default content
}
```

**Comment by Andrew-Chen-Wang** [comment]
> I still couldn't reproduce - I see this: in /template-nextjs/node_modules/next/image-types/global.d.ts which i think it providing the types - not sure why you don't have it.

Interesting. I do as well locally in my node_modules. I wonder why the CI wouldn't (again, this error only appears in CI). Since there really isn't much other way to fix it besides building, I'll close for now.
